File: code/utils/district_names.R
## Canonical Bangladesh district name harmonization (GADM / BBS / ERA5)

BBS_COXS_BAZAR <- "Cox\u2019s bazar"

name_recode <- c(
  "bandarban"     = "banderban",
  "bogra"         = "bogura",
  "brahamanbaria" = "brahmmanbaria",
  "chittagong"    = "chattogram",
  "comilla"       = "cumilla",
  "cox's bazar"   = "coxs bazar",
  "coxs bazar"    = "coxs bazar",
  "jessore"       = "jashore",
  "jhalokati"     = "jhalokathi",
  "khagrachhari"  = "khagrachari",
  "maulvibazar"   = "maulavi bazar",
  "nawabganj"     = "chapai nawabganj",
  "netrakona"     = "netrokona",
  "pirojpur"      = "perojpur"
)

normalize_apostrophes <- function(x) {
  x <- gsub("\u2019|\u2018|\u0060", "'", x, perl = TRUE)
  x
}

district_key <- function(x) {
  x <- normalize_apostrophes(trimws(tolower(as.character(x))))
  x <- gsub("['\u2019\u2018]", "", x, perl = TRUE)
  gsub("\\s+", " ", x)
}

clean_name <- function(x) district_key(x)

recode_shp <- function(x) {
  y <- district_key(x)
  mapped <- name_recode[y]
  ifelse(!is.na(mapped), mapped, y)
}

wb_to_bbs <- c(
  "Bandarban"     = "Banderban",
  "Bogra"         = "Bogura",
  "Brahamanbaria" = "Brahmmanbaria",
  "Chittagong"    = "Chattogram",
  "Comilla"       = "Cumilla",
  "Cox's Bazar"   = BBS_COXS_BAZAR,
  "Jessore"       = "Jashore",
  "Jhalokati"     = "Jhalokathi",
  "Khagrachhari"  = "Khagrachari",
  "Maulvibazar"   = "Maulavi Bazar",
  "Nawabganj"     = "Chapai Nawabganj",
  "Netrakona"     = "Netrokona",
  "Pirojpur"      = "Perojpur"
)

map_wb_district <- function(zila, bbs_districts = NULL) {
  if (!is.null(bbs_districts)) {
    cox <- grep("^Cox", bbs_districts, value = TRUE)
    if (length(cox) == 1L) wb_to_bbs["Cox's Bazar"] <- cox
  }
  out <- wb_to_bbs[zila]
  ifelse(is.na(out), zila, unname(out))
}

File: code/utils/district_names.py
"""Canonical Bangladesh district name harmonization (GADM / BBS / ERA5)."""

from __future__ import annotations

import re
import unicodedata

# Exact BBS spelling in wage and yield CSVs (curly apostrophe U+2019).
BBS_COXS_BAZAR = "Cox\u2019s bazar"

DISTRICT_MAPPING = {
    "Bandarban": "Banderban",
    "Bogra": "Bogura",
    "Brahamanbaria": "Brahmmanbaria",
    "Chittagong": "Chattogram",
    "Comilla": "Cumilla",
    "Cox'S Bazar": BBS_COXS_BAZAR,
    "Cox's Bazar": BBS_COXS_BAZAR,
    "Jessore": "Jashore",
    "Jhalokati": "Jhalokathi",
    "Khagrachhari": "Khagrachari",
    "Maulvibazar": "Maulavi Bazar",
    "Nawabganj": "Chapai Nawabganj",
    "Netrakona": "Netrokona",
    "Pirojpur": "Perojpur",
}


def normalize_apostrophes(name: str) -> str:
    """Replace Unicode apostrophe variants with ASCII for matching."""
    return (
        name.replace("\u2019", "'")
        .replace("\u2018", "'")
        .replace("\u0060", "'")
    )


def district_key(name) -> str | None:
    """Lowercase join key; strips apostrophes and collapses whitespace."""
    if name is None or (isinstance(name, float) and name != name):
        return None
    s = normalize_apostrophes(str(name).strip().lower())
    s = re.sub(r"['\u2019\u2018]", "", s)
    s = re.sub(r"\s+", " ", s)
    return s


def harmonize_district(name) -> str | None:
    """Map GADM / ERA5 names to BBS district labels."""
    if name is None or (isinstance(name, float) and name != name):
        return None
    s = str(name).strip()
    if s in DISTRICT_MAPPING:
        return DISTRICT_MAPPING[s]
    s_norm = normalize_apostrophes(s)
    for src, tgt in DISTRICT_MAPPING.items():
        if normalize_apostrophes(src) == s_norm:
            return tgt
    # Cox's Bazar: any apostrophe / case variant
    if district_key(s) == "coxs bazar":
        return BBS_COXS_BAZAR
    return s
